[GStreamer][WebCodecs][Debug] Assertions due to incorrect usage of WeakPtr

philn opened the pull request:
#### 54140a7b23efcb75955d08e7e7882371d1af3d99
<pre>
[GStreamer][WebCodecs][Debug] Assertions due to incorrect usage of WeakPtr
<a href="https://bugs.webkit.org/show_bug.cgi?id=264704">https://bugs.webkit.org/show_bug.cgi?id=264704</a>

Reviewed by Xabier Rodriguez-Calvar.

Adapt audio/video encoders to rely on ThreadSafeWeakPtr, that can be used from non-main threads.

* Source/WebCore/platform/audio/gstreamer/AudioEncoderGStreamer.cpp:
(WebCore::GStreamerAudioEncoder::encode):
(WebCore::GStreamerInternalAudioEncoder::GStreamerInternalAudioEncoder):
* Source/WebCore/platform/graphics/gstreamer/VideoEncoderGStreamer.cpp:
(WebCore::GStreamerVideoEncoder::encode):
(WebCore::GStreamerInternalVideoEncoder::GStreamerInternalVideoEncoder):

Canonical link: <a href="https://commits.webkit.org/270632@main">https://commits.webkit.org/270632@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/78d71e67bb0846f01c8e8fb53030f051bead8ba8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/25976 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4587 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27258 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28076 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23791 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/26294 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6349 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2015 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/23866 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26229 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3465 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22380 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28656 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/3085 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23333 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29397 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/23700 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23710 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27274 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3120 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1328 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4510 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6244 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3577 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3438 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

calvaris commented on Source/WebCore/platform/audio/gstreamer/AudioEncoderGStreamer.cpp:
```diff
@@ -266,12 +267,10 @@ GStreamerInternalAudioEncoder::GStreamerInternalAudioEncoder(const String& codec
+            if (protectedThis->m_isClosed)
                 return;
-            m_outputCallback({ WTFMove(encodedFrame) });
+            protectedThis->m_outputCallback({ WTFMove(encodedFrame) });
```
if you have a ref or a weakptr that you protect at the beginning of the lambda, you don't need to use the protector variable to access methods and attributes, as it was before.

calvaris commented on Source/WebCore/platform/graphics/gstreamer/VideoEncoderGStreamer.cpp:
```diff
@@ -210,14 +213,14 @@ GStreamerInternalVideoEncoder::GStreamerInternalVideoEncoder(const String& codec
-    m_harness = GStreamerElementHarness::create(WTFMove(element), [weakThis = WeakPtr { *this }, this](auto&, const GRefPtr<GstBuffer>& outputBuffer) {
-        if (!weakThis)
+    m_harness = GStreamerElementHarness::create(WTFMove(element), [weakThis = ThreadSafeWeakPtr { *this }, this](auto&, const GRefPtr<GstBuffer>& outputBuffer) {
+        if (!weakThis.get())
```
You don't need the get, do you?

philn replied:
I do, I see no bool operator in ThreadSafeWeakPtr.

philn committed:
[GStreamer][WebCodecs][Debug] Assertions due to incorrect usage of WeakPtr
https://bugs.webkit.org/show_bug.cgi?id=264704

Reviewed by Xabier Rodriguez-Calvar.

Adapt audio/video encoders to rely on ThreadSafeWeakPtr, that can be used from non-main threads.

* Source/WebCore/platform/audio/gstreamer/AudioEncoderGStreamer.cpp:
(WebCore::GStreamerAudioEncoder::encode):
(WebCore::GStreamerInternalAudioEncoder::GStreamerInternalAudioEncoder):
* Source/WebCore/platform/graphics/gstreamer/VideoEncoderGStreamer.cpp:
(WebCore::GStreamerVideoEncoder::encode):
(WebCore::GStreamerInternalVideoEncoder::GStreamerInternalVideoEncoder):

Canonical link: https://commits.webkit.org/270632@main
